internal/db/data_flow.md

```diff
@@ -0,0 +1,544 @@
+
+The `loadComposites()` function recursively traverses the block DAG:
+- Starts from the new block received from a peer
+- Walks backwards through parent blocks until reaching:
+  - Already processed blocks (present in current heads)
+  - Genesis blocks (no parents)
+  - Blocks with priority lower than the target
+- **Priority handling**:
+  - If new block priority ≥ target priority: Add to front of processing queue
+  - If new block priority < target priority: Create new merge target and recurse with parent blocks
+- This ensures blocks are processed in the correct order while handling divergent branches
+
+#### 3. Processing Blocks Recursively
+
+The merge processor handles blocks through several phases:
+
+**Decryption Phase:**
+- Checks if block is encrypted
+- Attempts to fetch encryption block from local store
+- On failure: Caches the failure and continues (will retry later)
+- On success: Decrypts and processes the block
+
+**CRDT Merge Phase:**
+- Initializes appropriate CRDT type based on field type
+- Applies the delta to merge changes
+- Handles conflicts according to CRDT semantics (e.g., LWW uses timestamp)
+
+**Recursive Processing:**
+- After processing a composite block, recursively processes all linked field blocks
+- Maintains processing order to ensure consistency
+
+#### 4. Encryption Key Recovery
+
+**Key Files:**
+- `internal/kms/pubsub.go` - KMS event handling
+- `internal/encryption/context.go` - Encryption context management
+
+If decryption fails during merge:
+- Publishes `encryption.RequestKeysEvent`
+- KMS subscribers attempt to fetch keys from the network
+- On key retrieval, calls `tryFetchMissingBlocksAndMerge()` which:
+  - Re-attempts the merge with newly available keys
+  - Can trigger another full merge cycle if successful
+
+#### 5. Final Updates
+
+After successful merge:
+- Updates secondary indexes with changed documents
+- Updates headstore with new head CIDs
+- Publishes `event.MergeComplete` for downstream consumers
+
+---
+
+## Part 4: System Behaviors
+
+This section covers cross-cutting concerns that affect the entire data flow.
+
+## Error Handling and Recovery
+
+DefraDB implements robust error handling throughout the data flow:
+
+### Key Recovery Mechanisms
+
+1. **Encryption Key Recovery**:
+   - Caches decryption failures
+   - Asynchronously requests keys from network
+   - Retries merge after key acquisition
+
+2. **Network Resilience**:
+   - Automatic retry for failed synchronizations
+   - Peer discovery for alternative sources
+   - Eventual consistency guarantees
+
+3. **Transaction Rollback**:
+   - Failed operations don't corrupt state
+   - Atomic updates ensure consistency
+   - Clear error propagation to clients
+
+## Summary
+
+The DefraDB data flow represents a sophisticated distributed system that orchestrates multiple components to provide a decentralized database experience.
+
+### Design Benefits
+
+This architecture enables DefraDB to:
```

Comment:
thought: The architecture does not really enable these things, the code does.  This section feels like marketing fluff/self-congratulations and encourages readers to ignore them and devalues the rest of the document.

You could perhaps keep these points but present them as a feature/requirement set of Defra.